### DISCUSSION TITLE
Wait for channel to be open after updating secret

### DISCUSSION THREAD

**Discussion by sidjakinsboriss** [ag2ai/faststream#1219] [Q&A]
Hi, I'm using the RabbitMQ OAuth 2.0 authentication backend, and therefore connecting to the broker with access tokens. These tokens can expire, so I have a wrapper around the publishing function that looks somewhat like this:

```python
def refresh_token(func):
    @wraps(func)
    async def wrapper(rabbit_client, *args, **kwargs):
        try:
            await func(rabbit_client, *args, **kwargs)
        except ChannelAccessRefused:
            new_token = rabbit_client.get_access_token()
            await rabbit_client.connection.update_secret(new_token, reason='Token expired')
            await func(rabbit_client, *args, **kwargs)

    return wrapper
```
The issue is that when calling `func` after updating the secret, a `ChannelInvalidStateError` is raised, since the channel was closed. Adding `await asyncio.sleep(0.01)` before trying to call `func` again fixes the issue, but seems a bit hacky. Is there some mechanism for waiting until the channel is open again?

**Comment by Lancetnik** [comment]
Which object type `rabbit_client` is?

**Comment by Lancetnik** [comment]
Well, I think, you are talking about `aio_pika.RobustConnection` update_secret method. Looks like it is blocking and should works without any sleep...

Unfortunately, I have never worked with RMQ OAuth2 and can't help you. Probably, you should ask for help in [**aio-pika**](https://github.com/mosquito/aio-pika/issues) original repo by creating Issue with your question.

**Reply by sidjakinsboriss** [reply]
The method itself works, but it seems like it doesn't wait for the channel to be re-opened or something. But I'll ask in aio-pika repo, thanks! :)
